cpp/src/arrow/acero/asof_join_node_test.cc

```diff
@@ -1770,5 +1771,53 @@ TEST(AsofJoinTest, DestroyNonStartedAsofJoinNode) {
       DeclarationToStatus(std::move(sink)));
 }
 
+// Reproduction of GH-46224: Hang when all left timestamps are greater than right
+// timestamps.
+TEST(AsofJoinTest, OneSideTsAllGreaterThanTheOther) {
```

Comment:
Just one thing: I see that it takes around 8s on a regular debug build (on my machine anyway). Should we make sure that it's not too costly under Valgrind?

Reply:
It was costly indeed. For the record: sanitizer time increases from 43.16s to 98.33s, valgrind time increases from 58.15s to 162.90s.

I reduced the iteration for sanitizer and valgrind to one in the latest commit.

Reply:
And the execution time remains about the same as existing (sanitizer: 44.11s vs 43.16s, valgrind: 59.68s vs 58.15s).

Reply:
Great, thanks!